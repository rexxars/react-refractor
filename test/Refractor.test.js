const React = require('react')
const ReactDOM = require('react-dom/server')
const js = require('refractor/lang/javascript')
const haml = require('refractor/lang/haml')
const Refractor = require('../src/Refractor')

beforeAll(() => {
  Refractor.registerLanguage(js)
  Refractor.registerLanguage(haml)
})

test('should render empty if no code is given', () => {
  expect(render({value: '', language: 'js'}, {withWrapper: true})).toBe(
    '<pre class="refractor language-js"><code class="language-js"></code></pre>'
  )
})

test('should render simple JS snippet correct', () => {
  expect(render({value: '"use strict";', language: 'js'}, {withWrapper: true})).toBe(
    '<pre class="refractor language-js">' +
      '<code class="language-js">' +
      '<span class="token string">&quot;use strict&quot;</span>' +
      '<span class="token punctuation">;</span>' +
      '</code>' +
      '</pre>'
  )
})

test('should use the specified language', () => {
  expect(render({value: '', language: 'haml'}, {withWrapper: true})).toBe(
    '<pre class="refractor language-haml"><code class="language-haml"></code></pre>'
  )
})

test('should be able to render inline', () => {
  expect(
    render(
      {value: 'var foo = "bar"', language: 'js', inline: true, className: 'moop'},
      {withWrapper: true}
    )
  ).toMatchSnapshot()
})

test('should be able to highlight specific lines with markers', () => {
  const language = 'javascript'
  const code = '{\n  title: "Sanity",\n  url: "https://sanity.io/"\n}\n'
  const markers = [2, {line: 3, className: 'url'}]
  expect(render({value: code, markers, language}, {withWrapper: true})).toMatchSnapshot()
})

function render(props, options) {
  const opts = options || {}
  const html = opts.reactAttrs
    ? ReactDOM.renderToString(React.createElement(Refractor, props))
    : ReactDOM.renderToStaticMarkup(React.createElement(Refractor, props))

  if (!opts.withWrapper) {
    return html.replace(/.*?<code.*?>([\s\S]*)<\/code>.*/g, '$1')
  }

  return html
}
