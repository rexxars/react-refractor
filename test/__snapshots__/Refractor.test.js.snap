// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not crash on markers beyond the number of lines in source 1`] = `"<div class=\\"refractor-marker\\"><span class=\\"token keyword\\">import</span> foo <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">&#x27;bar&#x27;</span></div>"`;

exports[`should be able to highlight specific lines with markers 1`] = `
"<pre class=\\"refractor language-javascript\\"><code class=\\"language-javascript\\"><span class=\\"token punctuation\\">{</span>
<div class=\\"refractor-marker\\">  <span class=\\"token literal-property property\\">title</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&quot;Sanity&quot;</span><span class=\\"token punctuation\\">,</span>
</div><div class=\\"url\\">  <span class=\\"token literal-property property\\">url</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&quot;https://sanity.io/&quot;</span>
</div><span class=\\"token punctuation\\">}</span>
</code></pre>"
`;

exports[`should be able to highlight specific, out-of-order lines with markers 1`] = `
"<pre class=\\"refractor language-javascript\\"><code class=\\"language-javascript\\"><div class=\\"refractor-marker\\"><span class=\\"token keyword\\">import</span> client <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">&#x27;part:@sanity/base/client&#x27;</span>
</div>
<span class=\\"token keyword\\">export</span> <span class=\\"token keyword\\">default</span> <span class=\\"token punctuation\\">{</span>
  <span class=\\"token literal-property property\\">name</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;post&#x27;</span><span class=\\"token punctuation\\">,</span>
  <span class=\\"token literal-property property\\">type</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;document&#x27;</span><span class=\\"token punctuation\\">,</span>
  <span class=\\"token literal-property property\\">title</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;Blog post&#x27;</span><span class=\\"token punctuation\\">,</span>
  <span class=\\"token function-variable function\\">initialValue</span><span class=\\"token operator\\">:</span> <span class=\\"token keyword\\">async</span> <span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">)</span> <span class=\\"token operator\\">=&gt;</span> <span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">{</span>
    <span class=\\"token literal-property property\\">publishedAt</span><span class=\\"token operator\\">:</span> <span class=\\"token keyword\\">new</span> <span class=\\"token class-name\\">Date</span><span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">.</span><span class=\\"token function\\">toISOString</span><span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">,</span>
<div class=\\"refractor-marker\\">    <span class=\\"token literal-property property\\">authors</span><span class=\\"token operator\\">:</span> <span class=\\"token keyword\\">await</span> client<span class=\\"token punctuation\\">.</span><span class=\\"token function\\">fetch</span><span class=\\"token punctuation\\">(</span><span class=\\"token template-string\\"><span class=\\"token template-punctuation string\\">\`</span><span class=\\"token string\\">
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">      *[_type == &quot;author&quot; &amp;&amp; &quot;marketing&quot; in responsbilities]{
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">        &quot;_type&quot;: &quot;authorReference&quot;,
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">        &quot;author&quot;: {
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">          &quot;_type&quot;: &quot;reference&quot;,
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">          &quot;_ref&quot;: _id
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">        }
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">      }
</span></span></div><div class=\\"refractor-marker\\"><span class=\\"token template-string\\"><span class=\\"token string\\">    </span><span class=\\"token template-punctuation string\\">\`</span></span><span class=\\"token punctuation\\">)</span>
</div><div class=\\"refractor-marker\\">  <span class=\\"token punctuation\\">}</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">,</span>
</div>  <span class=\\"token literal-property property\\">fields</span><span class=\\"token operator\\">:</span> <span class=\\"token punctuation\\">[</span>
    <span class=\\"token punctuation\\">{</span>
      <span class=\\"token literal-property property\\">name</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;title&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">type</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;string&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">title</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;Title&#x27;</span>
    <span class=\\"token punctuation\\">}</span><span class=\\"token punctuation\\">,</span>
    <span class=\\"token punctuation\\">{</span>
      <span class=\\"token literal-property property\\">name</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;slug&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">type</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;slug&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">title</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;Slug&#x27;</span>
    <span class=\\"token punctuation\\">}</span><span class=\\"token punctuation\\">,</span>
    <span class=\\"token punctuation\\">{</span>
      <span class=\\"token literal-property property\\">name</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;publishedAt&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">type</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;datetime&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">title</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;Published at&#x27;</span>
    <span class=\\"token punctuation\\">}</span><span class=\\"token punctuation\\">,</span>
    <span class=\\"token punctuation\\">{</span>
      <span class=\\"token literal-property property\\">name</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;authors&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">type</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;array&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token literal-property property\\">title</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;Authors&#x27;</span><span class=\\"token punctuation\\">,</span>
      <span class=\\"token keyword\\">of</span><span class=\\"token operator\\">:</span> <span class=\\"token punctuation\\">[</span>
        <span class=\\"token punctuation\\">{</span>
          <span class=\\"token literal-property property\\">type</span><span class=\\"token operator\\">:</span> <span class=\\"token string\\">&#x27;authorReference&#x27;</span>
        <span class=\\"token punctuation\\">}</span>  
      <span class=\\"token punctuation\\">]</span>
    <span class=\\"token punctuation\\">}</span>
    <span class=\\"token comment\\">// ...</span>
  <span class=\\"token punctuation\\">]</span>
<span class=\\"token punctuation\\">}</span></code></pre>"
`;

exports[`should be able to render inline 1`] = `"<code class=\\"moop\\" style=\\"display:inline\\"><span class=\\"token keyword\\">var</span> foo <span class=\\"token operator\\">=</span> <span class=\\"token string\\">&quot;bar&quot;</span></code>"`;
